core/blockchain.go

```diff
@@ -1042,11 +1048,29 @@ func (bc *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain [
 
 				// Always keep genesis block in active database.
 				if b.NumberU64() != 0 {
-					deleted = append(deleted, b)
+					deleted = append(deleted, &numberHash{b.NumberU64(), b.Hash()})
 				}
 				if time.Since(logged) > 8*time.Second {
 					log.Info("Migrating ancient blocks", "count", count, "elapsed", common.PrettyDuration(time.Since(start)))
 					logged = time.Now()
+					// Don't collect too much in-memory
+
+					// Sync the ancient store explicitly to ensure all data has been flushed to disk.
+					if err := bc.db.Sync(); err != nil {
+						return 0, err
+					}
+					// Wipe out canonical block data.
+					for _, nh := range deleted {
+						rawdb.DeleteBlockWithoutNumber(batch, nh.hash, nh.number)
+						rawdb.DeleteCanonicalHash(batch, nh.number)
+					}
+					// Wipe out side chain too.
+					for _, nh := range deleted {
+						for _, hash := range rawdb.ReadAllHashes(bc.db, nh.number) {
+							rawdb.DeleteBlock(batch, hash, nh.number)
+						}
+					}
+					deleted = deleted[0:]
```

Comment:
I don't really like this. We should no tie database batch sizes to log time triggers. The log is a simple progress report. If we want to batch the deletions, lets batch at 100K blocks (3.2MB worth of hashes) or something.

Reply:
Yup. 
Another more serious issue is that this erroneously writes the deletions to the `batch` but doesn't actually write the batch

Reply:
I fixed both these things three weeks ago, but apparently forgot to push. Anyway, pushed now :)